vcr/oidc4vci/service/identifiers.go

```diff
@@ -0,0 +1,164 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package service
+
+import (
+	"crypto/tls"
+	"fmt"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/vcr/log"
+	"github.com/nuts-foundation/nuts-node/vcr/oidc4vci"
+	"github.com/nuts-foundation/nuts-node/vdr/didservice"
+	"github.com/nuts-foundation/nuts-node/vdr/types"
+	"net/http"
+	"net/url"
+	"sync/atomic"
+	"time"
+)
+
+// IdentifierResolver defines the interface for resolving OpenID4VCI identifiers (of wallet and issuer).
+// The identifier is the base URL of the issuer or wallet, at which well-known endpoints can be found.
+type IdentifierResolver interface {
+	Resolve(id did.DID) (string, error)
+}
+
+var _ IdentifierResolver = DIDIdentifierResolver{}
+
+// DIDIdentifierResolver is a IdentifierResolver that uses resolves identifiers from DID documents.
+type DIDIdentifierResolver struct {
+	ServiceResolver didservice.ServiceResolver
+}
+
+func (i DIDIdentifierResolver) Resolve(id did.DID) (string, error) {
+	service, err := i.ServiceResolver.Resolve(didservice.MakeServiceReference(id, types.BaseURLServiceType), didservice.DefaultMaxServiceReferenceDepth)
+	if didservice.IsFunctionalResolveError(err) {
+		return "", nil
+	} else if err != nil {
+		return "", fmt.Errorf("unable to resolve %s service: %w", types.BaseURLServiceType, err)
+	}
+	var result string
+	_ = service.UnmarshalServiceEndpoint(&result)
+	if result != "" {
+		result = core.JoinURLPaths(result, "n2n", "identity", url.PathEscape(id.String()))
+	}
+	return result, nil
+}
+
+// NewTLSIdentifierResolver creates a IdentifierResolver that tries to derive the identifier from the TLS certificate if it can't be resolved using the DID document.
+// It does so by constructing the identifier from the CommonName and SubjectAlternativeNames of the certificate and requesting metadata.
+func NewTLSIdentifierResolver(underlying IdentifierResolver, config *tls.Config) IdentifierResolver {
+	result := tlsIdentifierResolver{
+		underlying:       underlying,
+		config:           config,
+		cachedIdentifier: new(atomic.Pointer[string]),
+		lastAttempt:      new(atomic.Pointer[time.Time]),
+	}
+	result.lastAttempt.Store(new(time.Time))
+	return result
+}
+
+const tlsAttemptInterval = time.Minute
+
+var tlsIdentifierResolverPort = 443
+
+func SetTLSIdentifierResolverPort(port int) {
+	tlsIdentifierResolverPort = port
+}
+
+var _ IdentifierResolver = tlsIdentifierResolver{}
+
+type tlsIdentifierResolver struct {
+	underlying       IdentifierResolver
```

Comment:
This pass-through resolver will become obsolete when all DID Documents are migrated?

Reply:
Now I think of it, `TLSIdentifierResolver` shouldn't ideally look into the DID document, but just try the TLS certificate. And then we introduce a `LocalIdentifierResolver` which tries `DIDIdentifierResolver` first, then `TLSIdentifierResolver` if it fails. But that might be a bit too overengineerd (thus, leave it like this)?